ykrt/src/compile/jitc_yk/codegen/x64/lsregalloc.rs

```diff
@@ -1074,38 +1100,46 @@ impl LSRegAlloc<'_> {
     /// Spill the value stored in `reg` if it is both (1) used in the future and (2) not already
     /// spilled. This updates `self.spills` (if necessary) but not `self.gp_reg_state` or
     /// `self.gp_regset`.
+    ///
+    /// Note: this function can change the CPU flags.
     fn spill_gp_if_not_already(&mut self, asm: &mut Assembler, cur_iidx: InstIdx, reg: Rq) {
-        match self.gp_reg_states[usize::from(reg.code())] {
+        match &self.gp_reg_states[usize::from(reg.code())] {
             RegState::Reserved => unreachable!(),
             RegState::Empty => (),
             RegState::FromConst(_, _) => (),
-            RegState::FromInst(query_iidx, _) => {
-                if self
-                    .rev_an
-                    .is_inst_var_still_used_after(cur_iidx, query_iidx)
-                    && self.spills[usize::from(query_iidx)] == SpillState::Empty
-                {
-                    let inst = self.m.inst(query_iidx);
-                    let bitw = inst.def_bitw(self.m);
-                    let bytew = inst.def_byte_size(self.m);
-                    debug_assert!(usize::try_from(bitw).unwrap() >= bytew);
-                    self.stack.align(bytew);
-                    let frame_off = self.stack.grow(bytew);
-                    let off = i32::try_from(frame_off).unwrap();
-                    match bitw {
-                        1 => {
-                            let old_ext =
-                                self.align_extensions(asm, reg, RegExtension::ZeroExtended);
-                            dynasm!(asm; mov BYTE [rbp - off], Rb(reg.code()));
-                            self.align_extensions(asm, reg, old_ext);
+            RegState::FromInst(query_iidxs, ext) => {
+                for query_iidx in query_iidxs {
+                    if self
+                        .rev_an
+                        .is_inst_var_still_used_after(cur_iidx, *query_iidx)
+                        && self.spills[usize::from(*query_iidx)] == SpillState::Empty
+                    {
+                        let inst = self.m.inst(*query_iidx);
+                        let bitw = inst.def_bitw(self.m);
+                        let bytew = inst.def_byte_size(self.m);
+                        debug_assert!(usize::try_from(bitw).unwrap() >= bytew);
+                        self.stack.align(bytew);
+                        let frame_off = self.stack.grow(bytew);
+                        let off = i32::try_from(frame_off).unwrap();
+                        match bitw {
+                            1 => {
+                                if *ext == RegExtension::ZeroExtended {
+                                    dynasm!(asm; mov BYTE [rbp - off], Rb(reg.code()));
+                                } else {
+                                    dynasm!(asm
+                                        ; bt Rd(reg.code()), 0
```

Comment:
As you mentioned in the commit message, this changes EFLAGS.

Are there situations where instr N+1 relies on the flag assignment of instr N? If so, are you worried that the flag clobbering could introduce hard to track down bugs?

I've been skimming the intel manual again. There are the `pushf*` instructions which save the flags to the stack, but it'd be expensive to do that conservatively (just in case the following instruction requires the flags), and I'm not sure if the architecture of the allocator/codegen allows it to peek ahead and ask "does someone need that flag assignment later?".

Tricky.

Reply:
IMHO no-one should rely on CPU flags being held across `assign_regs` calls, and I've adjusted the once instance of that I know of.

Reply:
OK, cool.

If the docstring for `assign_regs` doesn't already mention that you can't rely on flag assignments, I think that'd be a good addition.